### PR TITLE
kill -9 the ruby depselector on timeout

### PR DESCRIPTION
**Pull request by sdelano**
Timeouts on the ruby depselector have recently caused
zombie ruby processes to stick around that are not
able to tell that stdout has been closed. This patch
captures the pid of the ruby process at startup
and then sends a kill -9 to the pid when the depselector
process returns a timeout response.

### CODE REVIEW

**Comment by marcparadise** [comment]
Let's use this to get the pid instead: 

`{os_pid, OsPid} = erlang:port_info(Port, os_pid)`

File: src/chef_depsolver_worker.erl [hunk 1]
```diff
@@ -140,6 +157,14 @@ handle_call({solve, AllVersions, EnvConstraints, Cookbooks, Timeout}, _From, #st
     %% worst case scenario, we will double the receive timeout and add a
     %% small buffer of 50ms to account for inter-process communication
     %% time.
+    %%
+    %% If we do reach the Erlang-level receive timeout, then we want to
+    %% force-kill the ruby process. We handle the erlang timeout with
+    %% a forve-kill because there is a chance that the ruby process is
```
Comment (seth) [inline src/chef_depsolver_worker.erl]: force-kill

**Comment by seth** [comment]
+1 on @marcparadise's suggestion to capture the pid at the Erlang level with `port_info/2`.

I wonder if part of the problem is that pooler currently does a brutal kill of a member when given the `fail` atom on return member. Though I would have expected the process links to shutdown the port process in a reasonable fashion. That may not be the thing.

File: src/chef_depsolver_worker.erl [hunk 2]
```diff
@@ -140,6 +157,14 @@ handle_call({solve, AllVersions, EnvConstraints, Cookbooks, Timeout}, _From, #st
...
+    %% hung and will not be able to respond to stdout closing and clean
+    %% itself up. If, instead, the ruby process returns the timeout message,
+    %% we know that it remains in a state to respond to further requests
+    %% and therefore will be able to clean up after itself.
     ReceiveTimeout = (Timeout * 2) + 50,
```
Comment (seth) [inline src/chef_depsolver_worker.erl:162]: Why is this the right outer level timeout? My concern is that if this is too large, then we limp along more than we need to. Would `Timeout + X` be enough? Where X is either say 50-100?

Comment (sdelano) [inline src/chef_depsolver_worker.erl:162]: The "double the timeout" is there to give the ruby depsolver enough time to potentially return detailed information on a failed solve.

**Comment by sdelano** [comment]
@marcparadise Yes, let's just do the simple thing with `erlang:port_info`.

**Pull request merged**